Argmax and straight-through estimator

pmichel31415 opened the pull request:
Addresses #1199 

Only the argmax is done for now. Not sure this is very optimized (I'm basically looping over the batch dimension), I couldn't find a way to do complicated indexing in Eigen. Suggestions are welcome.

TODO:

- [x] Straight-through estimator
- [x] Python bindings

pmichel31415 commented:
Should be ready to merge once the tests are finished

neubig commented:
Is this finished? It looks like we're still missing some documentation for the argument `d`. I'll take a look when it's ready.

pmichel31415 commented:
OK I changed the API according to our email discussion.

I removed the "d" argument because it's not supported yet. I left it in the node because we might need it if we ever want to implement argmax along arbitrary dimensions.
